pkg/chunk/cache/memcached_client.go

```diff
@@ -92,17 +94,28 @@ func NewMemcachedClient(cfg MemcachedClientConfig, name string, r prometheus.Reg
 	client.Timeout = cfg.Timeout
 	client.MaxIdleConns = cfg.MaxIdleConns
 
-	dnsProviderRegisterer := prometheus.WrapRegistererWithPrefix("cortex_", prometheus.WrapRegistererWith(prometheus.Labels{
-		"name": name,
-	}, r))
+	// Check if we have a provider already. If not, create one.
+	// TODO(gouthamve): This is ugly. But short of not registering metrics, I don't see a good way.
+	// The main reason is that we create a cache _per schema_, which means we try to register the same name (chunks, index-cache-{read, write})
+	// several times, once per schema. Now this means that we cannot have different chunks cache for different schemas, but thats okay, _for now._
+	provider, ok := providersForName[name]
+	if !ok {
+		dnsProviderRegisterer := prometheus.WrapRegistererWithPrefix("cortex_", prometheus.WrapRegistererWith(prometheus.Labels{
+			"name": name,
+		}, r))
+
+		provider = dns.NewProvider(util.Logger, dnsProviderRegisterer, dns.GolangResolverType)
+
+		providersForName[name] = provider
+	}
```

Comment:
I think this belongs outside of this function. Perhaps caller can maintain the map of dns providers, and pass the correct one to the new memcached client?